File: credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    printf("Enter your credit card number ");
    long long cardnum = get_long_long();
    while (cardnum < 0)
    {
        printf("Please enter another card number ");
        cardnum = get_long_long();
    }
    long long digits;
    int count;
    count = 0;
    digits = cardnum;
    while (digits > 0)
    {
        digits = digits / 10;
        count++;
    }
    if (count != 13 && count != 15 && count != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    long long step1;
    int sum1, num;
    sum1 = 0;
    step1 = cardnum;
    num = 0;
    while (step1 > 0)
    {
        num = (step1 / 10) % 10;
        num = num * 2;
        if (num >= 10)
        {
            num = (num % 10) + (num / 10);
        }
        sum1 = sum1 + num;
        step1 = step1 / 100;
    }
    long long step12;
    int sum2, sum;
    step12 = cardnum / 10;
    sum2 = cardnum % 10;
    num = 0;
    while (step12 > 0)
    {
        num = (step12 / 10) % 10;
        sum2 = sum2 + num;
        step12 = step12 / 100;
    }
    sum = sum1 + sum2;
    long long type;
    if (sum % 10 == 0)
    {
        if (count == 15)
        {
            for (int i = 0; i < 13; i++)
            {
                cardnum = cardnum / 10;
            }
            if (cardnum == 34 || cardnum == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (count == 13)
        {
            for (int j = 0; j < 12; j++)
            {
                cardnum = cardnum / 10;
            }
            if (cardnum == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (count == 16)
        {
            type = cardnum;
            for (int k = 0; k < 15; k++)
            {
                type = type / 10;
            }
            if (type == 4)
            {
                printf("VISA\n");
            }
            else
            {
                type = cardnum;
                for (int x = 0; x < 14; x++)
                {
                    type = type / 10;
                }
                if (type == 51 || type == 52 || type == 53 || type == 54 || type == 55)
                {
                    printf("MASTERCARD\n");
                }
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: credit.py
from cs50 import get_int


def main():
    cc_no = get_int("Enter your credit card number:")

    checksum = 0
    first_two_digits = 0
    mult_by_two = false
    cc_no_length = 0

    AMEX_1 = 34
    AMEX_2 = 37
    AMEX_LENGTH = 15

    FIRST_MC = 51
    LAST_MC = 55
    MC_LENGTH = 16

    FIRST_VISA = 40
    LAST_VISA = 49
    VISA_LENGTH1 = 13
    VISA_LENGTH2 = 16

    # iterate over digits
    while (cc_no > 0):
        # for each digit, starting from the last:
        cur_digit = cc_no % 10

        # add the new digit to the checksum, doubling if necessary
        if (mult_by_two):
            # accounting for the case where need to add digits together
            if cur_digit < 5:
                checksum += (cur_digit * 2)
            else:
                checksum += (cur_digit * 2 - 9)
        else:
            checksum += cur_digit

        # keep track of beginning of number to identify credit card company
        first_two_digits = cur_digit * 10 + (first_two_digits / 10)

        # update state and drop last digit
        mult_by_two = not mult_by_two
        cc_no = cc_no / 10
        cc_no_length = cc_no_length + 1

    # validate checksum, company identifier, and number length
    if (checksum % 10):
        print("INVALID")
    elif (((first_two_digits == AMEX_1) or (first_two_digits == AMEX_2)) and (cc_no_length == AMEX_LENGTH)):
        print("AMEX")
    elif ((first_two_digits >= FIRST_VISA) and (first_two_digits <= LAST_VISA) and ((cc_no_length == VISA_LENGTH1) or (cc_no_length == VISA_LENGTH2))):
        print("VISA")
    elif ((first_two_digits >= FIRST_MC) and (first_two_digits <= LAST_MC) and (cc_no_length == MC_LENGTH)):
        print("MASTERCARD")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()